library/init/category/result.lean

```diff
@@ -0,0 +1,56 @@
+prelude
+
+import init.logic init.category.monad init.category.alternative
+
+universe u
+
+inductive result (E A : Type u)
+| error {} : E → result
+| ok {} : A → result
+
+def result.return {E A : Type u} (a : A) : result E A :=
+result.ok a
+
+def result.map {E A B : Type u} (f : A → B) : result E A → result E B
+| (result.error err) := result.error err
+| (result.ok v) := result.ok $ f v
+
+def result.bind {E A B : Type u} (ma : result E A) (f : A → result E B) : result E B :=
+match ma with
+| (result.error err) := result.error err
+| (result.ok v) := f v
+end
+
+instance (E : Type u) : monad (result E) :=
+{ pure := @result.return E ,
+  map := @result.map E,
+  bind := @result.bind E }
+
+@[reducible] def except_t (M : Type u → Type u) (E A : Type u) :=
+M (result E A)
+
+section except_t
+
+parameter (M : Type u → Type u)
+parameter [monad_M : monad M]
+parameter E : Type u
```

Comment:
We usually use
```lean
variable (m : Type u → Type u)
variable [monad m]
```

BTW, why is the output universe and input universes the same?